Fix crash where node or node->name are NULL

aviposs opened the pull request:
This can happen when an unknown tag was encountered in wbxml_tag_get_xml_name.

bellmich commented:
Hi, 

do you have a test or test document which causes a crash? First I would like to add a test which crashes the library.

Best regards, 
Michael

aviposs commented:
I wrote this to fix a situation that I encountered in the wild.

This is all the information I was able to glean from the logs:
1. I had wbxml which I converted to a tree and encountered an unknown tag which caused the node->name to be null. (I suspect that this was a new Activesync 16.0 tag, the libwbxml version which was running doesn't support the 16.0 tags)
2. I ran wbxml_tree_to_wbxml on the tree which was created in the previous step and when the code which I fixed was reached, it crashed because node->name was null.

I don't have a test case, but I suspect that if you create a test which runs wbxml_tree_to_wbxml on a tree which has tags which don't exist in wbxml_tables.c, the issue will be reproduced. This is a realistic scenario, as the protocol is updated often enough.

bellmich commented:
I generated a WBXML from an ActiveSync XML document. Then I converted this WBXML document to XML but with SyncML as the language. This results in several <unknown> tags but there was no crash. Which parameters did you use to parse the WBXML document?

aviposs commented:
The wbxml tree was created with:
wbxml_tree_from_wbxml(dataPtr, dataSize, WBXML_LANG_ACTIVESYNC, WBXML_CHARSET_UTF_8, wbxmlTreePtr)

While creating this tree, in the wbxml_tag_get_xml_name function, it returned WBXML_ELT_UNKNOWN_NAME for one of the tags and then I tried to re-convert the tree back to data using wbxml_tree_to_wbxml with the following parameters changed:
    WBXMLGenWBXMLParams params;
    params.wbxml_version = WBXML_VERSION_13;
    params.keep_ignorable_ws = TRUE;
    params.use_strtbl = FALSE;
    params.produce_anonymous = TRUE;
When running this, the code crashed on a null ptr at the point that I changed.

bellmich commented:
Which version of libwbxml do you use?

aviposs commented:
In wbxml_config.h:
#define WBXML_LIB_VERSION "0.11.2"

bellmich commented:
I have implemented a test which generate a WBXML document from an Active Sync XML document. This document I converted back to XML but with SyncML as the specified language. I used the parameters which you defined above. I get a XML document with several <unknown> elements but there was no crash.

You wrote about the tree functions. The tree functions are not part of the public API. Do you use only the public API or do you use the private functions of the library too (in your normal code)?
